### PR TITLE
fix `make test:all` failures seen on macOS Tahoe

### PR DESCRIPTION
**Pull request by scottlamb**
## Description

`make test:all` fails with a variety of errors on macOS Tahoe, I think mostly relating to it using a newer/stricter clang (21) than CI and, presumably, most contributor's machines. Each commit here fixes one class of failure (and has an example of the prior compiler error in the commit description).

As you can tell I let my AI agent write the changes, but I reviewed them and think they're reasonable.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by scottlamb** [comment]
Hmm, I see [this lint formatter failure](https://github.com/qmk/qmk_firmware/actions/runs/24166261061/job/70551619465?pr=26136)...but the failure is on lines not changed in my PR. It looks like the file hasn't changed since the lint was set up, and the lint only checks changed files. Would you like me to add a commit to the beginning that updates the formatting of the files my PR touches?

**Comment by zvecr** [comment]
Looks like our formatting workflow had a slight oversight for '.hpp` files -> #26137.

~~Once merged, and the subsequent formatting PRs merged, it shouldn't be an issue.~~ Should be resolve since 18ed7c6cafec5a14751500ebfbb75d1262cfefbe, if you want to update the branch for this PR.

File: tests/test_common/keyboard_report_util.hpp [hunk 1]
```diff
@@ -35,5 +35,8 @@ class KeyboardReportMatcher : public testing::MatcherInterface<report_keyboard_t
 
 template<typename... Ts>
 inline testing::Matcher<report_keyboard_t&> KeyboardReport(Ts... keys) {
-    return testing::MakeMatcher(new KeyboardReportMatcher(std::vector<uint8_t>({keys...})));
+    std::vector<uint8_t> v;
+    v.reserve(sizeof...(keys));
+    (v.push_back(keys), ...);
```
Comment (zvecr) [inline tests/test_common/keyboard_report_util.hpp]: This fold expression would be c++17, where we currently target c++14.

Comment (scottlamb) [reply tests/test_common/keyboard_report_util.hpp]: Switched to an explicit cast. In hindsight this is much clearer anyway.

Purely out of curiosity, why C++14? I'm certainly used to embedded projects having to use crappy vendor toolchains for arch support, but I think the tests only run on the host anyway?

**Comment by scottlamb** [comment]
> Should be resolve since https://github.com/qmk/qmk_firmware/commit/18ed7c6cafec5a14751500ebfbb75d1262cfefbe, if you want to update the branch for this PR.

Rebased, thanks!

File: tests/test_common/keyboard_report_util.hpp (hunk 1 above)
Comment (drashna) [reply tests/test_common/keyboard_report_util.hpp]: I was curious myself. This PR actually covers a bunch of that. https://github.com/qmk/qmk_firmware/pull/17114

But basically, it's due to compiler support on various platforms (and things like debian/ubuntu only recently updated the avr/arm gcc compilers from the 5.x version). 

However, the qmk-toolchain support means that we do have more control over the compiler version....

File: quantum/unicode/unicode.c [hunk 2]
```diff
@@ -178,7 +178,7 @@ void set_unicode_input_mode(uint8_t mode) {
 }
 
 static void cycle_unicode_input_mode(int8_t offset) {
-#if UNICODE_SELECTED_MODES != -1
+#ifdef UNICODE_SELECTED_MODES_DEFINED
```
Comment (zvecr) [inline quantum/unicode/unicode.c]:
```suggestion
#ifdef UNICODE_SELECTED_MODES
```

File: quantum/unicode/unicode.c [hunk 3]
```diff
@@ -136,7 +136,7 @@ static void unicode_play_song(uint8_t mode) {
 
 void unicode_input_mode_init(void) {
     eeconfig_read_unicode_mode(&unicode_config);
-#if UNICODE_SELECTED_MODES != -1
+#ifdef UNICODE_SELECTED_MODES_DEFINED
```
Comment (zvecr) [inline quantum/unicode/unicode.c]:
```suggestion
#ifdef UNICODE_SELECTED_MODES
```

File: quantum/unicode/unicode.c [hunk 4]
```diff
@@ -66,7 +66,7 @@ unicode_config_t unicode_config;
 uint8_t          unicode_saved_mods;
 led_t            unicode_saved_led_state;
 
-#if UNICODE_SELECTED_MODES != -1
+#ifdef UNICODE_SELECTED_MODES_DEFINED
```
Comment (zvecr) [inline quantum/unicode/unicode.c]:
```suggestion
#ifdef UNICODE_SELECTED_MODES
```

File: quantum/unicode/unicode.c [hunk 5]
```diff
@@ -48,8 +48,8 @@
 
 // Comma-delimited, ordered list of input modes selected for use (e.g. in cycle)
 // Example: #define UNICODE_SELECTED_MODES UNICODE_MODE_WINCOMPOSE, UNICODE_MODE_LINUX
```
Comment (zvecr) [inline quantum/unicode/unicode.c:50]:
```suggestion
```

File: quantum/unicode/unicode.c [hunk 6]
```diff
@@ -48,8 +48,8 @@
...
-#ifndef UNICODE_SELECTED_MODES
-#    define UNICODE_SELECTED_MODES -1
+#ifdef UNICODE_SELECTED_MODES
+#    define UNICODE_SELECTED_MODES_DEFINED
 #endif
```
Comment (zvecr) [inline quantum/unicode/unicode.c]: `UNICODE_SELECTED_MODES_DEFINED` doesn't add enough value over just checking for the define directly
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: tests/test_common/keyboard_report_util.hpp (hunk 1 above)
Comment (zvecr) [reply tests/test_common/keyboard_report_util.hpp]: The intention is more to keep things in step. Tests still bring in the core code being asserted. Allowing tests be a different version could lead to inconsistencies being introduced into core.

`qmk-toolchain` doesn't directly cover the case of the compiler used for tests, though it does influence a  bump to standards when asserting what our minimum could become.

File: quantum/unicode/unicode.c (hunk 6 above)
Comment (scottlamb) [reply quantum/unicode/unicode.c]: Fair point; squashed these into their commit (now bc6a7d9b1b), and updated the doc to fix the `-1` mention.

**Comment by zvecr** [comment]
Apart from the additional commits that have snuck in, and the [unresolved suggestion](https://github.com/qmk/qmk_firmware/pull/26136/changes#r3078779374) removal of an out of date comment, LGTM.

File: quantum/unicode/unicode.c (hunk 5 above)
Comment (scottlamb) [reply quantum/unicode/unicode.c]: Should I really remove this? I didn't view it as stale: if you remove the `// Example: ` prefix, that `#define` should work.

**Review by scottlamb — COMMENTED** [review]
> Apart from the additional commits that have snuck in

Oops, should be fixed. (I must have rebased on top of `develop` (stale local branch) instead of an up-to-date `origin/develop` and it cherry-picked `origin/develop` commits, changing their hashes.)

File: quantum/unicode/unicode.c (hunk 5 above)
Comment (zvecr) [reply quantum/unicode/unicode.c]: No one should be modifying core to enable this behaviour. The examples in the docs should serve as the example users follow.

 It could potentially be relocated above https://github.com/scottlamb/qmk_firmware/blob/develop/quantum/unicode/unicode.c#L70.

Comment (scottlamb) [reply quantum/unicode/unicode.c]: Fair enough, removed.

**Review by drashna — APPROVED** [review]

**Review by zvecr — APPROVED** [review]
